File: app/src/main/java/com/example/todocloud/fragment/ReminderDatePickerDialogFragment.java
package com.example.todocloud.fragment;

import android.app.DatePickerDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatDialogFragment;
import android.widget.DatePicker;

import com.example.todocloud.R;

import java.util.Calendar;
import java.util.Date;

public class ReminderDatePickerDialogFragment extends AppCompatDialogFragment implements
    DatePickerDialog.OnDateSetListener {

  private int year;
  private int month;
  private int day;
  private Date date;
  private Calendar calendar = Calendar.getInstance();
  private IReminderDatePickerDialogFragment listener;

  @Override
  public void onAttach(Context context) {
    super.onAttach(context);
    listener = (IReminderDatePickerDialogFragment) getTargetFragment();
  }

  @NonNull
  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {
    date = (Date) getArguments().get("reminderDate");
    calendar.setTime(date);
    year = calendar.get(Calendar.YEAR);
    month = calendar.get(Calendar.MONTH);
    day = calendar.get(Calendar.DAY_OF_MONTH);
    DatePickerDialog datePickerDialog = new DatePickerDialog(
        getActivity(), R.style.MyPickerDialogTheme, this, year, month, day
    );
    prepareDatePickerDialogButtons(datePickerDialog);
    return datePickerDialog;
  }

  private void prepareDatePickerDialogButtons(DatePickerDialog datePickerDialog) {
    datePickerDialog.setButton(
        DialogInterface.BUTTON_POSITIVE,
        getString(R.string.btnTime),
        datePickerDialog
    );
    datePickerDialog.setButton(
        DialogInterface.BUTTON_NEGATIVE,
        getString(R.string.btnDeleteReminder),
        new DialogInterface.OnClickListener() {

          @Override
          public void onClick(DialogInterface dialog, int which) {
            listener.onDeleteReminder();
          }

        }
    );
  }

  @Override
  public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
    calendar.set(year, monthOfYear, dayOfMonth);
    date.setTime(calendar.getTimeInMillis());
    listener.onSelectReminderDate(date);
    dismiss();
  }

  public interface IReminderDatePickerDialogFragment {
    void onSelectReminderDate(Date date);
    void onDeleteReminder();
  }

}

File: app/src/main/java/com/rolandvitezhu/todocloud/ui/activity/main/dialogfragment/ReminderDatePickerDialogFragment.kt
package com.rolandvitezhu.todocloud.ui.activity.main.dialogfragment

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.widget.DatePicker
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.lifecycle.ViewModelProvider
import com.rolandvitezhu.todocloud.R
import com.rolandvitezhu.todocloud.ui.activity.main.fragment.CreateTodoFragment
import com.rolandvitezhu.todocloud.ui.activity.main.fragment.ModifyTodoFragment
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.TodosViewModel
import org.threeten.bp.LocalDateTime

class ReminderDatePickerDialogFragment : AppCompatDialogFragment(), OnDateSetListener {

    /*private var year = 0
    private var month = 0
    private var day = 0
    private var date: LocalDateTime? = null*/
    private val todosViewModel by lazy {
        ViewModelProvider(requireActivity()).get(TodosViewModel::class.java)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        /*val thereIsNoDateTime = requireArguments()[Constant.REMINDER_DATE_TIME] == null
        date =
                if (thereIsNoDateTime)
                    LocalDateTime.now()
                else
                    requireArguments()[Constant.REMINDER_DATE_TIME] as LocalDateTime?*/
        if (todosViewModel.ldtReminderDateTime == null)
            todosViewModel.ldtReminderDateTime = LocalDateTime.now()

        val year = todosViewModel.ldtReminderDateTime?.year ?: 0
        val month = todosViewModel.ldtReminderDateTime?.monthValue ?: 0
        val day = todosViewModel.ldtReminderDateTime?.dayOfMonth ?: 0

        val datePickerDialog = DatePickerDialog(
                requireActivity(),
                R.style.MyPickerDialogTheme,
                this,
                year, month - 1, day
        )
        prepareDatePickerDialogButtons(datePickerDialog)

        return datePickerDialog
    }

    private fun prepareDatePickerDialogButtons(datePickerDialog: DatePickerDialog) {
        datePickerDialog.setButton(
                DialogInterface.BUTTON_POSITIVE,
                getString(R.string.reminderdatepicker_positivebuttontext),
                datePickerDialog/*Message.obtain()*/
        )
    }

    override fun onDateSet(view: DatePicker, year: Int, month: Int, day: Int) {
        /*date = LocalDateTime.of(year, month + 1, day, date?.hour as Int, date?.minute as Int)*/
        todosViewModel.ldtReminderDateTime =
                LocalDateTime.of(
                        year,
                        month + 1,
                        day,
                        todosViewModel.ldtReminderDateTime?.hour ?: 0,
                        todosViewModel.ldtReminderDateTime?.minute ?: 0)

        if (targetFragment is CreateTodoFragment?)
            /*date?.let { */(targetFragment as CreateTodoFragment?)?.onSelectReminderDate(/*it*/)/* }*/
        else if (targetFragment is ModifyTodoFragment?)
            /*todosViewModel.ldtReminderDateTime?.let {*/
                (targetFragment as ModifyTodoFragment?)?.onSelectReminderDate(/*it*/)/* }*/    // NEXT

        dismiss()
    }
}